plugins/formal-verification/src/org/jetbrains/kotlin/formver/embeddings/ExpEmbedding.kt

```diff
@@ -0,0 +1,187 @@
+/*
+ * Copyright 2010-2023 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.formver.embeddings
+
+import org.jetbrains.kotlin.formver.domains.NullableDomain
+import org.jetbrains.kotlin.formver.domains.UnitDomain
+import org.jetbrains.kotlin.formver.scala.silicon.ast.*
+import org.jetbrains.kotlin.formver.scala.toScalaBigInt
+import viper.silver.ast.DomainType
+import javax.sound.sampled.LineEvent.Type
+
+sealed interface ExpEmbedding {
+    val type: TypeEmbedding
+    val viperExp: Exp
+
+    fun withType(newType: TypeEmbedding): ExpEmbedding =
+        when {
+            type == newType -> this
+            type !is NullableTypeEmbedding && newType is NullableTypeEmbedding -> NullableOf(this.withType(newType.elementType))
+            // This NullLit conversion is just hard-coded here to not break the tests,
+            // it can be deleted once the to-do below has been solved.
+            this is NullLit && newType is NullableTypeEmbedding -> NullLit(newType)
+            type is NullableTypeEmbedding && (type as NullableTypeEmbedding).elementType.isSubTypeOf(newType) ->
+                ValOfNullable(this).withType(newType)
+            type is NullableTypeEmbedding && newType is NullableTypeEmbedding -> TODO("Add domain function in the Nullable domain to convert from one type of nullable to another if their element types are subtypes.")
+            type.isSubTypeOf(newType) -> throw NotImplementedError("Type $type is a subtype of $newType but no conversion function was specified.")
+            else -> throw IllegalArgumentException("Expression $this of type $type is not a subtype of $newType and can thus not be converted to it.")
+        }
+}
+
+fun <E : ExpEmbedding> List<E>.viperExps(): List<Exp> = map { it.viperExp }
+
+data class Add(val left: ExpEmbedding, val right: ExpEmbedding) : ExpEmbedding {
```

Comment:
With things done this way, there's no room for position information and such anymore.  I think that a separate commit that gets rid of it all would be a good idea.